File: pm_core/src/infrastructure/store/adapter/work_item_adapter.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package adapter

import (
	"context"
	"fmt"

	"github.com/serp/pm-core/src/core/domain/entity"
	"github.com/serp/pm-core/src/core/port/store"
	"github.com/serp/pm-core/src/infrastructure/store/mapper"
	"github.com/serp/pm-core/src/infrastructure/store/model"
	"gorm.io/gorm"
)

type WorkItemAdapter struct {
	db     *gorm.DB
	mapper *mapper.WorkItemMapper
}

func NewWorkItemAdapter(db *gorm.DB) store.IWorkItemPort {
	return &WorkItemAdapter{
		db:     db,
		mapper: mapper.NewWorkItemMapper(),
	}
}

func (a *WorkItemAdapter) CreateWorkItem(ctx context.Context, tx *gorm.DB, workItem *entity.WorkItemEntity) (*entity.WorkItemEntity, error) {
	db := a.getDB(tx)
	workItemModel := a.mapper.ToModel(workItem)
	if err := db.WithContext(ctx).Create(workItemModel).Error; err != nil {
		return nil, fmt.Errorf("failed to create work item: %w", err)
	}
	return a.mapper.ToEntity(workItemModel), nil
}

func (a *WorkItemAdapter) GetWorkItemByID(ctx context.Context, id int64) (*entity.WorkItemEntity, error) {
	var workItemModel model.WorkItemModel
	if err := a.db.WithContext(ctx).First(&workItemModel, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil
		}
		return nil, fmt.Errorf("failed to get work item by id: %w", err)
	}
	return a.mapper.ToEntity(&workItemModel), nil
}

func (a *WorkItemAdapter) UpdateWorkItem(ctx context.Context, tx *gorm.DB, workItem *entity.WorkItemEntity) error {
	db := a.getDB(tx)
	workItemModel := a.mapper.ToModel(workItem)
	if err := db.WithContext(ctx).Save(workItemModel).Error; err != nil {
		return fmt.Errorf("failed to update work item: %w", err)
	}
	return nil
}

func (a *WorkItemAdapter) SoftDeleteWorkItem(ctx context.Context, tx *gorm.DB, workItemID int64) error {
	db := a.getDB(tx)
	if err := db.WithContext(ctx).Model(&model.WorkItemModel{}).
		Where("id = ?", workItemID).
		Update("active_status", "DELETED").Error; err != nil {
		return fmt.Errorf("failed to soft delete work item: %w", err)
	}
	return nil
}

func (a *WorkItemAdapter) getDB(tx *gorm.DB) *gorm.DB {
	if tx != nil {
		return tx
	}
	return a.db
}

File: pm_core/src/main/java/serp/project/pmcore/infrastructure/store/adapter/WorkItemAdapter.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.pmcore.infrastructure.store.adapter;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.util.Pair;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.stereotype.Component;
import serp.project.pmcore.core.domain.dto.filter.WorkItemFilterRequest;
import serp.project.pmcore.core.domain.entity.WorkItemEntity;
import serp.project.pmcore.core.port.store.IWorkItemPort;
import serp.project.pmcore.infrastructure.store.mapper.WorkItemMapper;
import serp.project.pmcore.infrastructure.store.mapper.WorkItemRowMapper;
import serp.project.pmcore.infrastructure.store.query.WorkItemQueryBuilder;
import serp.project.pmcore.infrastructure.store.repository.IWorkItemRepository;

import java.util.List;
import java.util.Optional;

@Component
@RequiredArgsConstructor
@Slf4j
public class WorkItemAdapter implements IWorkItemPort {

    private final IWorkItemRepository workItemRepository;
    private final WorkItemMapper workItemMapper;

    private final NamedParameterJdbcTemplate jdbcTemplate;
    private final WorkItemQueryBuilder queryBuilder;
    private final WorkItemRowMapper rowMapper;

    @Override
    public WorkItemEntity saveWorkItem(WorkItemEntity workItem) {
        return workItemMapper.toEntity(
                workItemRepository.save(workItemMapper.toModel(workItem))
        );
    }

    @Override
    public Optional<WorkItemEntity> getWorkItemById(Long id, Long tenantId) {
        return workItemRepository.findByIdAndTenantId(id, tenantId)
                .map(workItemMapper::toEntity);
    }

    @Override
    public List<WorkItemEntity> getWorkItemsByProjectId(Long projectId, Long tenantId) {
        return workItemMapper.toEntities(
                workItemRepository.findAllByTenantIdAndProjectId(tenantId, projectId)
        );
    }

    @Override
    public List<WorkItemEntity> getWorkItemsByIssueTypeId(Long issueTypeId, Long tenantId) {
        return workItemMapper.toEntities(
                workItemRepository.findAllByTenantIdAndIssueTypeId(tenantId, issueTypeId)
        );
    }

    @Override
    public void deleteWorkItemById(Long id, Long tenantId) {
        workItemRepository.deleteByIdAndTenantId(id, tenantId);
    }

    @Override
    public Pair<List<WorkItemEntity>, Long> searchWorkItems(Long tenantId, WorkItemFilterRequest filter) {
        var qr = queryBuilder.build(tenantId, filter);
        log.debug("WorkItem search SQL: {}", qr.dataSql());
        log.debug("WorkItem count SQL: {}", qr.countSql());

        List<WorkItemEntity> data = jdbcTemplate.query(qr.dataSql(), qr.params(), rowMapper);
        Long total = jdbcTemplate.queryForObject(qr.countSql(), qr.params(), Long.class);
        return Pair.of(data, total);
    }
}
